Integrate XNNPACK with custom class for packing weights.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34547 Enable threading for XNNPACK ops.
* #34319 Pass to remove prepacking ops.
* #34048 JIT pass to insert XNNPACK ops
* **#34047 Integrate XNNPACK with custom class for packing weights.**

Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

kimishpatel committed:
Integrate XNNPACK with custom class for packing weights.
Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/xnnpack/Factory.cpp:
```diff
@@ -8,11 +8,16 @@ namespace native {
     const caffe2::TypeMeta dtype,
     const c10::MemoryFormat memory_format) {
-  static GuardingAllocator<0u, XNN_EXTRA_BYTES> allocator;
+  auto* allocator_ptr = get_guarding_allocator();
```
Just curious - can we ever arrive here from different threads now that we are using custom op registration?  I am not sure if the allocator is thread safe.

kimishpatel replied:
Thats a good point. Thanks for pointing it out. But I think this is ok, because allocator is not keeping track of any data that my introduce race?

AshkanAliabadi commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -775,10 +775,6 @@
 
-- func: _conv2d_prepack(Tensor weight, Tensor? bias=None, int[2] stride=1, int[2] padding=0, int[2] dilation=1, int groups=1, float? output_min=None, float? output_max=None) -> Tensor
-
-- func: _conv2d_packed(Tensor packed_weight, Tensor input) -> Tensor
```
@gchanan @dzhulgakov @ezyang Are we OK with this?  I am no sure which way we decided to go eventually, but want to make sure we are not closing any doors for GPU in case there is benefit to pre-packing there as well (i.e. we have to do direct registration like this change is doing for GPU as well.)

kimishpatel replied:
Ashkan, so we had the internal chat group discussion about this and my understanding is that we decided to go the way quantization is doing. To be fair, I did try to keep this change, however to make codegen work with custom class type will involve changes in quite a few places which become quite invasive. For custom class types I feel the current way is safer without exposing the custom type to broader source code. But if you guys have a different opinion, we can discuss.

ezyang replied:
Not to spread FUD, but do you know if torchbind works on mobile?

jamesr66a replied:
@ezyang after my stack of patches https://github.com/pytorch/pytorch/pull/34098 it should work fine on mobile

AshkanAliabadi commented on aten/src/ATen/native/xnnpack/OpContext.cpp:
```diff
@@ -0,0 +1,59 @@
+
+c10::intrusive_ptr<XNNPackLinearOpContext>
+XNNPackLinearOpContext::create_context(at::Tensor&& weight,
+    c10::optional<at::Tensor>&& bias,
```
rvalue references cannot bind to lvalues by design.  The accepted paradigm is to either (1) pass parameters by value, (2) pass by [const] lvalue reference if copies are expensive, or (3) provide a (const lvalue reference, rvalue reference) pair overload.  Providing an rvalue reference only interface limits use.

By the way, tensors in PyTorch are ref-counted wrappers to internal implementations so they are cheap to shallow-copy around.

kimishpatel replied:
> Providing an rvalue reference only interface limits use.

So this was by intention. I wanted to make it explicit that callsite has to provide rvalue reference. Doing by value introduces shallow copy that are ref counted, and ref counting is not exactly cheap, right? So I thought this made more sense. 
But I might be mistaken, so do provide more context if you disagree and I can learn something new.
We can also do options 3 that you suggested, but I wanted to make sure I provide the context for my reasoning.

ngimel replied:
Agreed, shallow ref-counted copy is not cheap.

AshkanAliabadi replied:
I think the question is whether you envision a case where create_context is called with weight and bias tensors that are neither temporaries nor moved-from objects?  In other words, do we want to be able to  support something like this:

```
Tensor weight;  // Some weight tensor somewhere
create_context( std::move(weight) );
weight.blah();  // weight is moved from.  It can no longer be used so this is not valid.
```

kimishpatel replied:
Yes, so I did not envision a former usecase. In my mind weight is always moved from while creating context and thus not valid after that.

AshkanAliabadi replied:
Sounds good, thanks!

AshkanAliabadi commented on aten/src/ATen/native/xnnpack/OpContext.cpp:
```diff
@@ -0,0 +1,59 @@
+    c10::optional<at::Tensor>&& bias,
+    std::vector<int64_t>&& padding,
+    std::vector<int64_t>&& stride,
+    std::vector<int64_t>&& dilation,
```
Likewise all here and other places in code potentially with the exception of Operator which is just a non-copyable std::unique_ptr.  I would personally just opt for option 1 or 2 above 95%+ of the time.  Even 3 is overkill unless one is writing a generic library.

kimishpatel replied:
As I mentioned before, passing by value maybe little expensive, but I suppose create_context is called only once at load time so it might be ok. If you this seems more clean, I am ok with it, but as I said, I wanted to explain my reasoning and see if I might be wrong in that. Then something new to learn :).

AshkanAliabadi commented on aten/src/ATen/native/xnnpack/OpContext.h:
```diff
@@ -0,0 +1,97 @@
+  private:
+    Tensor orig_weight_;
+    c10::optional<Tensor> orig_bias_;
+    Context op_context_;
```
OK, so if I am reading the code correctly, it seems we need separate contexts for Linear and Convolution anyway.  I think it would be better just to separate them, and expose them in Convolution.h and Linear.h as opposed to bundling them together in Common.h.  Or maybe I am missing something?

kimishpatel replied:
I explained the reasoning above, but you are right, we are storing context separately anyway.

AshkanAliabadi reviewed (commented):
Thanks for working on this Kimish!

kimishpatel commented on aten/src/ATen/native/xnnpack/Common.h:
```diff
@@ -9,6 +9,47 @@
+
+using Operator = std::unique_ptr<xnn_operator, Deleter>;
+
+struct Context final {
```
So the reason to expose this to Common.h is because the context has to be stored in the new custom class, e.g. LinearOpContext. It might be possible to move it back where it was but that would require custom class registration to be moved along with it, which may make code organization little unclean. 
Regarding, having separate Context per op: The reason for me to do it this was mainly because in XNNPACK there is a single operator context. I am open to both options, via separate op specific context as well as union approach, but I prefer to keep that in Common.h for the reasons mentioned above. Union approach might be preferable to be sort of in sync with what XNNPACK does?

kimishpatel commented on aten/src/ATen/native/xnnpack/Convolution.cpp:
```diff
@@ -228,6 +151,100 @@ Tensor create_and_run(
+	std::move(dilation),
+	groups,
+	{},
+	{});
```
I agree. My plan for ReLU or other clamping fusion would actually modify this API to pass along min/max. I did not do that as part of this PR. If you prefer it doing in this PR itself, I am ok with it, but since we are not yet doing fusion I did not yet expose it.

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/xnnpack/Convolution.h:
```diff
@@ -0,0 +1,49 @@
+    const c10::optional<Tensor>& bias,
+    const IntArrayRef padding_,
+    const IntArrayRef stride_,
+    const IntArrayRef dilation_,
```
Why do these have trailing underscores?

kimishpatel replied:
Thanks for pointing this out. I had just moved these functions around, so it was there before. I will remote trailing _.

dreiss commented on caffe2/utils/threadpool/pthreadpool_impl.cc:
```diff
@@ -28,7 +28,20 @@ void pthreadpool_compute_1d(
 
 size_t pthreadpool_get_threads_count(pthreadpool_t threadpool) {
-  return reinterpret_cast<caffe2::ThreadPool*>(threadpool)->getNumThreads();
+  if (threadpool) {
```
Can this be pulled into a separate diff that we can land now?

kimishpatel replied:
Here: https://github.com/pytorch/pytorch/pull/34087. Added you to it.

dreiss commented on test/test_xnnpack_integration.py:
```diff
@@ -0,0 +1,347 @@
+        torch.testing.assert_allclose(ref_result, output_linear_xnnpack, rtol=1e-2, atol=1e-3)
+
+
+    @given(batch_size=st.integers(1, 3),
```
Please be sure to test Linear and Conv with batch size = 0.  (It doesn't have to be part of this same general test.  Just test that you can run it and get a proper shape output.)

dreiss reviewed (commented):
Haven't had time to do a full review yet, but I wanted to leave these minor comments asap.

kimishpatel commented on aten/src/ATen/native/xnnpack/Common.h:
```diff
@@ -9,6 +9,47 @@
+  std::vector<int64_t> weight_size;
+  std::vector<int64_t> padding;
+  std::vector<int64_t> stride;
+  std::vector<int64_t> dilation;
```
My understanding is that, this is for private variables only.

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -0,0 +1,115 @@
+
+namespace {
+torch::jit::class_<XNNPackLinearOpContext> register_xnnpack_linear_op_context_class() {
+  static int custom_class_handler = torch::jit::register_custom_class_handler();
```
This is going away after https://github.com/pytorch/pytorch/pull/34098

jamesr66a commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -0,0 +1,115 @@
+}
+
+torch::jit::class_<XNNPackConv2dOpContext> register_xnnpack_conv2d_op_context_class() {
+  static int custom_class_handler = torch::jit::register_custom_class_handler();
```
Same here

jamesr66a reviewed (commented):
This looks fine to me (I only looked at TorchBind-related stuff), but I think you should wait for these PRs to land: https://github.com/pytorch/pytorch/pull/34098 (excluding top one) since it's changing how torchbind stuff internal to the framework works

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

ezyang commented:
I'm deferring to @jamesr66a for this PR review.

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

dzhulgakov reviewed (commented):
Looks nice!

One question I have is whether operator bindings should be backend-specific vs be generic "stateful" linear/conv ops with a hidden backend dispatch decision that doesn't get serialized.

It's a similar discussion to qengine that we had for quantized ops. Basic criteria is whether we want serialized model to explicitly refer to XNNPACK or just generically say "weights are prepacked" and let runtime pick the best engine. For gengine we went with "generic serilalization" so far.

Can XNNPACK be thought as generic engine as well? In that case we probably should bind custom classes without  'xnnpack' in the name and provide fallback that "does no packing" in case XNNPACK is absent (and the same path can be extended later if we have more prepackable engines).

@AshkanAliabadi @ngimel  @jamesr66a - what do you think?

kimishpatel commented:
@dzhulgakov, some responses below:
> One question I have is whether operator bindings should be backend-specific vs be generic "stateful" linear/conv ops with a hidden backend dispatch decision that doesn't get serialized.
> 
So for mobile specific the flow I imagine is as follows. You have a scripted/traced model `model`. We do something like:
```
torch._C.XNNPACKify...(model)
model.save_for_lite_interpreter(..)
```
Then the byte code is loaded and run on mobile. Since we dont have full JIT available on mobile, we have to generate backend specific bytecode. At least, this is what I imagine for now.

Ignoring this part, if we were to make this generic "stateful" linear/conv passes, then I imagine we would do something like
```
torch._C.backend_transform(model, backend='xnnpack')
model.save_for... if needed.
```
I guess that is fine to do and it helps hide complexity from the frontend (from where the graph rewriting pass is called), but for mobile case we will have to do this AOT anyway. 

> 
> Can XNNPACK be thought as generic engine as well? In that case we probably should bind custom classes without 'xnnpack' in the name and provide fallback that "does no packing" in case XNNPACK is absent (and the same path can be extended later if we have more prepackable engines).
> 

I think this is possible. We can think of GPU specific optimization as similar passes perhaps, although Ashkan will have better idea on that.
Not sure if we want to do this now or when we actually need it. I prefer doing it when need arises mainly because the changes here are not making you buy in a specific design that is hard to alter later. But do note that we wont be running this graph rewriting passes on lite interpreter, so all that backend specific optimization, xnnpack or GPU will have to AOT, I think. Correct me if I am wrong.

AshkanAliabadi commented:
> Looks nice!
> 
> One question I have is whether operator bindings should be backend-specific vs be generic "stateful" linear/conv ops with a hidden backend dispatch decision that doesn't get serialized.
> 
> It's a similar discussion to qengine that we had for quantized ops. Basic criteria is whether we want serialized model to explicitly refer to XNNPACK or just generically say "weights are prepacked" and let runtime pick the best engine. For gengine we went with "generic serilalization" so far.
> 
> Can XNNPACK be thought as generic engine as well? In that case we probably should bind custom classes without 'xnnpack' in the name and provide fallback that "does no packing" in case XNNPACK is absent (and the same path can be extended later if we have more prepackable engines).
> 
> @AshkanAliabadi @ngimel @jamesr66a - what do you think?

I don't think I have a complete understanding of the intricacies involved in each approach to be able to provide any meaningful input, so I defer to the wisdom of those who do.

dreiss commented on aten/src/ATen/native/xnnpack/Common.h:
```diff
@@ -9,6 +9,53 @@
+  ContextConv2D(Operator&& o, std::vector<int64_t> w_size,
+      std::vector<int64_t> padding_,
+      std::vector<int64_t> stride_,
+      std::vector<int64_t> dilation_) {
```
Naming is off here.  Members should have underscores, and argument names should match, the full member names, minus the underscores.  Formatting seems off as well.

dreiss replied:
Also, prefer member initialization syntax to assignment syntax.

kimishpatel replied:
> Naming is off here. Members should have underscores, and argument names should match, the full member names, minus the underscores. Formatting seems off as well.

My understanding is that, only private member names trail with _. Hence the change. I can reverse it so that member name trail with _. But that was the reason.

kimishpatel replied:
> Also, prefer member initialization syntax to assignment syntax.

Will do.

dreiss replied:
Oh, I didn't think of public members.  I would probably still use the underscore just to be consistent, but I don't feel strongly about it.

dreiss commented on aten/src/ATen/native/xnnpack/Common.h:
```diff
@@ -31,7 +78,7 @@ struct Layout final {
 
       // Handle the case where batch size is zero.
-      int64_t batch = std::max<int64_t>(1, tensor[0]);
+      int64_t batch = tensor[0];
```
Seems like this might be a bug in the old version.  But also, unrelated to this diff, is this function taking an argument called "tensor" that is meant to be a tensor shape?  That's confusing.

kimishpatel replied:
Yes, this was a bug in the older version. This is to handle batch size 0.

dreiss commented on aten/src/ATen/native/xnnpack/Convolution.cpp:
```diff
@@ -79,113 +67,48 @@ bool available(
          (c10::DeviceType::CPU == input.device().type()) &&
          (kFloat == input.scalar_type()) &&
-         (input.size(Layout::Activation4D::batch) > 0) &&
+         (input.size(Layout::Activation4D::batch) >= 0) &&
```
Please separate the batch 0 fixes from the custom class diff.

kimishpatel replied:
Sure, I will separate this out. Regarding your previous question, yes the naming is little confusing.

dreiss commented on aten/src/ATen/native/xnnpack/Convolution.cpp:
```diff
@@ -79,113 +67,48 @@ bool available(
   using namespace internal;
 
   const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+  const Tensor padded_input_nhwc = allocate_padded_if_needed(input_nhwc);
```
Should this be a separate diff as well?

kimishpatel replied:
Yes it is possible, however I prefer we do it here, mostly for convenience reason :). However if you want I can try to separate out.

dreiss commented on aten/src/ATen/native/xnnpack/OpContext.cpp:
```diff
@@ -0,0 +1,59 @@
+namespace xnnpack {
+
+c10::intrusive_ptr<XNNPackLinearOpContext>
+XNNPackLinearOpContext::create_context(at::Tensor&& weight,
```
Should have a newline after the open paren.

dreiss commented on aten/src/ATen/native/xnnpack/Convolution.cpp:
```diff
@@ -228,6 +151,100 @@ Tensor create_and_run(
+        std::move(dilation),
+        groups,
+        {},
+        {});
```
Do we need min/max here to fuse relu?

kimishpatel replied:
So my plan was to make the necessary changes, including ones needed in op API, in the PR for relu fusion.

dreiss commented on aten/src/ATen/native/xnnpack/OpContext.h:
```diff
@@ -0,0 +1,98 @@
+
+  public:
+    XNNPackLinearOpContext(Tensor&& weight,
+        c10::optional<Tensor>&& bias) {
```
Probably best to apply clang-format to all of the new files.

dreiss commented on aten/src/ATen/native/xnnpack/OpContext.h:
```diff
@@ -0,0 +1,98 @@
+    }
+    //XNNPackLinearOpContext(const XNNPackLinearOpContext&) = delete; // Copy constructor
+    //XNNPackLinearContext& operator=(const XNNPackLinearOpContext&) = delete; // Assign
+    //Move construct and move assign?
```
Clean up?

dreiss commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -0,0 +1,113 @@
+                std::tie(weight, bias) = op_context->unpack();
+                return std::make_tuple(
+                    std::move(weight),
+                    std::move(bias));
```
Seems like you're unpacking and repacking the tuple.  Why not just `return op_context->unpack();`?

kimishpatel replied:
Good catch. Will fix.

dreiss commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -0,0 +1,113 @@
+                    std::move(std::get<0>(state)),
+                    std::move(std::get<1>(state)),
+                    {},
+                    {}
```
Don't we need to preserve min/max?

kimishpatel replied:
Yes. Min, max will have to be saved as part of the context. Again my plan was to enable that path when we enable relu fusion.

dreiss commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -0,0 +1,113 @@
+
+// Op registeration
+static auto registry =
+  // Registering under xnnpack namespace for now. As we add more backend requiring similar functionality
```
I vote "_xnnpack" namespace for now.  We're likely to learn soon that we screwed this up.  Let's give ourselves a few free opportunities to fix before we lock in a name.

dzhulgakov replied:
I'd strongly insist to rename it before we ship any model with it. I'm pretty confident we should have generic wrappers for stateful conv/linear. Can we at least add a big TODO to fix it before we deploy any model to the wild?

dreiss commented on torch/csrc/Module.cpp:
```diff
@@ -560,6 +566,7 @@ static PyMethodDef TorchMethods[] = {
   {"_get_qengine", (PyCFunction)THPModule_qEngine, METH_NOARGS, nullptr},
   {"_set_qengine", (PyCFunction)THPModule_setQEngine, METH_O, nullptr},
   {"_supported_qengines", (PyCFunction)THPModule_supportedQEngines, METH_NOARGS, nullptr},
+  {"_is_xnnpack_enabled", (PyCFunction)THPModule_isEnabledXNNPACK, METH_NOARGS,     nullptr},
```
Extra spaces.

dreiss commented on torch/backends/xnnpack/__init__.py:
```diff
@@ -0,0 +1,25 @@
+
+# This is the sys.modules replacement trick, see
+# https://stackoverflow.com/questions/2447353/getattr-on-a-module/7668273#7668273
+sys.modules[__name__] = XNNPACKEngine(sys.modules[__name__], __name__)
```
I see that we do this for CUDA and quantization, but does anyone know why this is necessary?

kimishpatel commented on aten/src/ATen/native/xnnpack/Common.h:
```diff
@@ -9,6 +9,53 @@
+  int64_t output_channels;
+
+  ContextLinear() = default;
+
```
Good catch. I will tag `delete` to it.

kimishpatel commented on aten/src/ATen/native/xnnpack/Common.h:
```diff
@@ -9,6 +9,53 @@
+  std::vector<int64_t> weight_size;
+  std::vector<int64_t> padding;
+  std::vector<int64_t> stride;
+  std::vector<int64_t> dilation;
```
Will do.

kimishpatel replied:
BTW, these parameters are allocated only by the caller, they are to be moved from in this function. It was supposed to accept by rvalue ref as function argument instead by value. I will move to std::array nonetheless.

dreiss replied:
Maybe we can take IntArrayRef from the caller?

kimishpatel replied:
The calling site already had a vector created. We could have created IntArrayRef out of that, but that would have just done yet another construction of object. I thought moving was better. Anyhow, I just moved to std::array as you suggested. One less thing to dereference before getting to data.

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/xnnpack/Factory.cpp:
```diff
@@ -21,15 +26,28 @@ Tensor empty_with_tail_padding(
   return tensor.resize_(size, memory_format);
 }
 
+Tensor allocate_padded_if_needed(const Tensor& input_contig) {
```
@dzhulgakov, I think the current approach will work with views as well, since the guarding allocator guarantees some number of safe-to-access bytes beyond the end of the tensor (which is the requirement for xnnpack). However, I am not sure how to use `storage_offset + nbytes` and `storage().capacity()` to deduce that there are some  safe-to-access bytes allocated beyond the end of the tensor. I think only the allocator knows truly how many bytes it allocated. Please correct me if I am wrong.
Honestly, it would be great if we can determine 1) whether tensor dimensions are such that no out-of-bound access will occur. or 2) if guarding allocator is used. 2 we covered. But if we cover 1 as well, it will greatly simplify allocations and for the most use cases I think 1 should be satisfied.

dreiss replied:
I think we should stick with this version now because it is simple and safe.  We can follow up to discuss Dima's suggestion later.

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel commented:
@dreiss, addressed your comments.

dreiss commented:
> One question I have is whether operator bindings should be backend-specific vs be generic "stateful" linear/conv ops with a hidden backend dispatch decision that doesn't get serialized.

Sorry, I missed this question earlier.  In the very short term, I think we should move forward on this diff with the underscore namespace and not block any work on this decision.  Beyond that, I lean toward the "generic" approach.  I don't think this causes any problems for the light interpreter as long as we hide the distinction inside the ops, the same way we do with fbgemm/qnnpack.  One issue, though, is that XNNPACK supports fused RELU and RELU6, which we need to use to get maximum performance.  So if we make the ops generic, those need to become part of the interface.  That doesn't seem like too big of a deal for conv and linear, but what about when we want fused relu for add and maxpool. :/

dreiss reviewed (approved):
Test looks quite solid, which gives me the confidence to approve.  It seems like there are a lot of layers of context creation functions, and I got a bit lost tracing them.  Maybe we can discuss tomorrow and see if there is a way to simplify or at least add a comment with a short overview of the relationship.

kimishpatel commented:
> > One question I have is whether operator bindings should be backend-specific vs be generic "stateful" linear/conv ops with a hidden backend dispatch decision that doesn't get serialized.
> 
> Sorry, I missed this question earlier. In the very short term, I think we should move forward on this diff with the underscore namespace and not block any work on this decision. Beyond that, I lean toward the "generic" approach. I don't think this causes any problems for the light interpreter as long as we hide the distinction inside the ops, the same way we do with fbgemm/qnnpack. One issue, though, is that XNNPACK supports fused RELU and RELU6, which we need to use to get maximum performance. So if we make the ops generic, those need to become part of the interface. That doesn't seem like too big of a deal for conv and linear, but what about when we want fused relu for add and maxpool. :/

I see both yours and Dima's point. Yes it should be doable similar to quant stuff, except for sorting out all the fusion related issues. On the other hand, I had thought of `save_for_lite_interpreter(..., backend=...)` interface. But this goes against what Dima's concern was for quantization stuff as well. So it make sense to hide different backend behind a generic API and backend is brought in only at execution.

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

kimishpatel committed:
Update on "Integrate XNNPACK with custom class for packing weights."


Summary:
This PR integrates the added xnnpack conv2d and linear op via
custom class registration for packed weights. The packed struct
is serializable.

Test Plan:
python test test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20185657](https://our.internmc.facebook.com/intern/diff/D20185657)

[ghstack-poisoned]

dzhulgakov commented:
Just to explain the motivation, I think we should try to avoid backend-specifics leaking into user API or serialization format wherever possible. Not doing so (i.e. exposing xnnpack directly as user-visible primitive) has several drawbacks:
- adds to API surface area that user needs to be aware of
- reduces our ability to switch to better backends in the future. What if ZNNPACK comes out tomorrow and we want to switch silently existing models to it?
- externalizes to our users to care about existence of different backends - now serialized models are not runnable everywhere anymore

Instead of exposing directly the backend, we should distill the underlying primitives that are required for this backend to run efficiently. In this case - it's ability to rewrite weights for Conv and Linear. This primitive is also likely useful for a lot of other backends too (e.g. PackedFC in C2 on x86)

Sure, in some cases backend implementation might not allow seamless wrapping (TRT being extreme example), but at least we should try for the reasons listed above. For XNNPACK I think it's pretty doable.

As for ReLU fusion listed above, there's a simple solution: add a fused op, provide straightforward implementation for it in op library (i.e. call conv+relu directly), override this op for a more efficient implementation when xnnpack is available.

It's ok that we landed this PR, but I'd strongly insist on sorting this out before we start deploying XNNPACK models to prod.

kimishpatel commented:
> Just to explain the motivation, I think we should try to avoid backend-specifics leaking into user API or serialization format wherever possible. Not doing so (i.e. exposing xnnpack directly as user-visible primitive) has several drawbacks:
> 

@dzhulgakov, that sounds reasonable. If we want to have generic wrappers, then this should be done before we ship any models with this. Will add a TODO for the same. But I have some questions regarding your other inputs. Would appreciate your comments.

> * adds to API surface area that user needs to be aware of
> * reduces our ability to switch to better backends in the future. What if ZNNPACK comes out tomorrow and we want to switch silently existing models to it?
> * externalizes to our users to care about existence of different backends - now serialized models are not runnable everywhere anymore
> 

For 2, I see and completely agree with your point.

For 1 and 3, do you expect users to write models with say `linear_prepack` and `linear_packed` ops, as opposed to just `linear`, and then expect the runtime to remove `linear_prepack`? From user, meaning model engineer writing model or someone developing wrapper API on top of PT, perspective, I expect that to just say give me linear and optimization pass can choose to transform it or not to `linear_prepack` and `linear_packed`.

> Instead of exposing directly the backend, we should distill the underlying primitives that are required for this backend to run efficiently. In this case - it's ability to rewrite weights for Conv and Linear. This primitive is also likely useful for a lot of other backends too (e.g. PackedFC in C2 on x86)
> 
> Sure, in some cases backend implementation might not allow seamless wrapping (TRT being extreme example), but at least we should try for the reasons listed above. For XNNPACK I think it's pretty doable.
> 

Agreed.

> As for ReLU fusion listed above, there's a simple solution: add a fused op, provide straightforward implementation for it in op library (i.e. call conv+relu directly), override this op for a more efficient implementation when xnnpack is available.
> 

This also sounds reasonable, but I am assuming you dont mean that, that is the only way to expose relu fusion? Meaning, yes we can expose a fused op, but we can also have optimization pass that recognizes this pattern and does fusion, much like other JIT fusion passes. 

> It's ok that we landed this PR, but I'd strongly insist on sorting this out before we start deploying XNNPACK models to prod.

Let me try to set something up, to sort this out and all of us can be on the same page.
